File: desktop/src/features/workflows/agencyAutomations.ts
import { isTauri } from "@tauri-apps/api/core";

import {
  agencyRuntimeEndpoint,
  fetchAgencyRuntimeConfig,
} from "@/features/surfaces/lib/agencyRuntime";
import { invokeTauri } from "@/shared/api/tauri";

export type AgencyAutomation = {
  id: string;
  name: string;
  description: string;
  ownerAgent: string;
  scope: string;
  enabled: boolean;
  triggers: string[];
  runMode: "thread";
  canRun: boolean;
};

function normalizeAutomation(value: unknown): AgencyAutomation | null {
  if (typeof value !== "object" || value === null) return null;
  const source = value as Record<string, unknown>;
  const id = typeof source.id === "string" ? source.id.trim() : "";
  if (!id) return null;
  const runMode =
    source.runMode === "thread" || source.run_mode === "thread"
      ? "thread"
      : "thread";
  return {
    id,
    name:
      typeof source.name === "string" && source.name.trim()
        ? source.name.trim()
        : id,
    description:
      typeof source.description === "string" ? source.description.trim() : "",
    ownerAgent:
      typeof source.ownerAgent === "string"
        ? source.ownerAgent.trim()
        : typeof source.owner_agent === "string"
          ? source.owner_agent.trim()
          : "",
    scope: typeof source.scope === "string" ? source.scope.trim() : "",
    enabled: source.enabled !== false,
    triggers: Array.isArray(source.triggers)
      ? source.triggers.flatMap((trigger) =>
          typeof trigger === "string" && trigger.trim() ? [trigger.trim()] : [],
        )
      : [],
    runMode,
    canRun: source.canRun === true || source.can_run === true,
  };
}

export async function discoverAgencyAutomations(): Promise<AgencyAutomation[]> {
  try {
    const payload = isTauri()
      ? await invokeTauri<unknown>("discover_agency_automations")
      : await fetch(
          agencyRuntimeEndpoint(
            await fetchAgencyRuntimeConfig(),
            "/api/agency/automations",
          ),
        ).then((response) => {
          if (!response.ok) throw new Error(`HTTP ${response.status}`);
          return response.json() as Promise<unknown>;
        });
    const values = Array.isArray(payload)
      ? payload
      : typeof payload === "object" &&
          payload !== null &&
          Array.isArray((payload as Record<string, unknown>).automations)
        ? ((payload as Record<string, unknown>).automations as unknown[])
        : [];
    return values
      .flatMap((value) => {
        const automation = normalizeAutomation(value);
        return automation ? [automation] : [];
      })
      .sort((left, right) => {
        if (left.enabled !== right.enabled) return left.enabled ? -1 : 1;
        return left.name.localeCompare(right.name);
      });
  } catch {
    return [];
  }
}

File: desktop/src-tauri/src/commands/agency_automations.rs
use serde::{Deserialize, Serialize};
use std::time::Duration;
use tauri::{AppHandle, State};

use crate::app_state::AppState;
use crate::commands::agency_runtime_endpoint;

const AGENCY_AUTOMATION_DISCOVERY_TIMEOUT: Duration = Duration::from_secs(3);

#[derive(Debug, Deserialize, Serialize, PartialEq)]
#[serde(rename_all = "camelCase")]
pub struct AgencyAutomationSummary {
    id: String,
    name: String,
    #[serde(default)]
    description: String,
    #[serde(default, alias = "owner_agent")]
    owner_agent: String,
    #[serde(default)]
    scope: String,
    enabled: bool,
    #[serde(default)]
    triggers: Vec<String>,
    #[serde(default, alias = "run_mode")]
    run_mode: String,
    #[serde(default, alias = "can_run")]
    can_run: bool,
}

/// Discover provider-neutral automation definitions. Execution recipes remain
/// private to the Agency runtime; Buzz receives only enough metadata to render
/// the automation and its thread projection.
#[tauri::command]
pub async fn discover_agency_automations(
    app: AppHandle,
    state: State<'_, AppState>,
) -> Result<Vec<AgencyAutomationSummary>, String> {
    let response = state
        .media_fetch_client
        .get(agency_runtime_endpoint(&app, "/api/agency/automations")?)
        .timeout(AGENCY_AUTOMATION_DISCOVERY_TIMEOUT)
        .send()
        .await
        .map_err(|error| format!("Agency automation discovery failed: {error}"))?;
    if !response.status().is_success() {
        return Err(format!(
            "Agency automation discovery returned HTTP {}",
            response.status()
        ));
    }
    let payload = response
        .json::<serde_json::Value>()
        .await
        .map_err(|error| format!("Agency automation discovery response was invalid: {error}"))?;
    let payload = payload.get("automations").cloned().unwrap_or(payload);
    serde_json::from_value::<Vec<AgencyAutomationSummary>>(payload)
        .map_err(|error| format!("Agency automation discovery response was invalid: {error}"))
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn accepts_protocol_snake_case_and_serializes_for_frontend() {
        let automation = serde_json::from_value::<AgencyAutomationSummary>(serde_json::json!({
            "id": "founder-briefing",
            "name": "founder-briefing",
            "owner_agent": "weaver",
            "enabled": true,
            "triggers": ["cron", "manual"],
            "run_mode": "thread",
            "can_run": true
        }))
        .expect("automation");
        assert_eq!(automation.owner_agent, "weaver");
        assert_eq!(automation.run_mode, "thread");
        assert!(automation.can_run);
        let frontend = serde_json::to_value(automation).expect("frontend JSON");
        assert_eq!(frontend["ownerAgent"], "weaver");
        assert_eq!(frontend["runMode"], "thread");
    }
}
